controllers/bpfman-operator/application-programs.go

```diff
@@ -0,0 +1,184 @@
+*/
+
+package bpfmanoperator
+
+import (
+	"context"
+	"fmt"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	bpfmaniov1alpha1 "github.com/bpfman/bpfman-operator/apis/v1alpha1"
+	internal "github.com/bpfman/bpfman-operator/internal"
+)
+
+// BpfApplicationReconciler reconciles a BpfApplication object
+type BpfApplicationReconciler struct {
+	ReconcilerCommon
+}
+
+func (r *BpfApplicationReconciler) getRecCommon() *ReconcilerCommon {
+	return &r.ReconcilerCommon
+}
+
+func (r *BpfApplicationReconciler) getFinalizer() string {
+	return internal.BpfApplicationControllerFinalizer
+}
+
+//+kubebuilder:rbac:groups=bpfman.io,resources=bpfapplications,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=bpfman.io,resources=bpfapplications/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=bpfman.io,resources=bpfapplications/finalizers,verbs=update
+
+func (r *BpfApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Logger = log.FromContext(ctx)
+
+	appProgram := &bpfmaniov1alpha1.BpfApplication{}
+	if err := r.Get(ctx, req.NamespacedName, appProgram); err != nil {
+		// Reconcile was triggered by bpfProgram event, get parent appProgram Object.
+		if errors.IsNotFound(err) {
+			bpfProgram := &bpfmaniov1alpha1.BpfProgram{}
+			if err := r.Get(ctx, req.NamespacedName, bpfProgram); err != nil {
+				if errors.IsNotFound(err) {
+					r.Logger.V(1).Info("bpfProgram not found stale reconcile, exiting", "Name", req.NamespacedName)
+				} else {
+					r.Logger.Error(err, "failed getting bpfProgram Object", "Name", req.NamespacedName)
+				}
+				return ctrl.Result{}, nil
+			}
+
+			// Get owning appProgram object from ownerRef
+			ownerRef := metav1.GetControllerOf(bpfProgram)
+			if ownerRef == nil {
+				return ctrl.Result{Requeue: false}, fmt.Errorf("failed getting bpfProgram Object owner")
+			}
+
+			if err := r.Get(ctx, types.NamespacedName{Namespace: corev1.NamespaceAll, Name: ownerRef.Name}, appProgram); err != nil {
+				if errors.IsNotFound(err) {
+					r.Logger.Info("Application Programs from ownerRef not found stale reconcile exiting", "Name", req.NamespacedName)
+				} else {
+					r.Logger.Error(err, "failed getting Application Programs Object from ownerRef", "Name", req.NamespacedName)
+				}
+				return ctrl.Result{}, nil
+			}
+
+		} else {
+			r.Logger.Error(err, "failed getting Application Programs Object", "Name", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+	}
+
+	return r.reconcileAppPrograms(ctx, appProgram)
+}
+
+func (r *BpfApplicationReconciler) reconcileAppPrograms(ctx context.Context, application *bpfmaniov1alpha1.BpfApplication) (ctrl.Result, error) {
+	var result ctrl.Result
+	var err error
+
+	for _, prog := range application.Spec.Programs {
```

Comment:
My gut feeling is that we won't have to care about the individual programs type here.  We just need to find all of the bpfPrograms for which this applicationProgram is the parent.  Then, loop through them and get the status.  For the bpfman-agent controller, we'll need to look at each program type and do some initialization of a reconciler data structure for the given program type, and then call the bpfman-agent reconcileProgram() function.